native/native_main.py

```diff
@@ -14,43 +15,527 @@
+                powershell_cmd,
+                powershell_args,
+                restart_ps1_path,
+            )
+
+            startupinfo = subprocess.STARTUPINFO()
+            startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
+
+            subprocess.check_output(
+                [
+                    "powershell",
+                    "-NonInteractive",
+                    "-NoProfile",
+                    "-WindowStyle",
+                    "Minimized",
+                    "-InputFormat",
+                    "None",
+                    "-ExecutionPolicy",
+                    "Bypass",
+                    "-Command",
+                    "Register-ScheduledTask \
+                    -TaskName '%s' \
+                    -Force \
+                    -Action (New-ScheduledTaskAction \
+                              -Execute '%s' \
+                              -Argument '%s') \
+                    -Trigger (New-ScheduledTaskTrigger \
+                              -Once \
+                              -At \
+                (Get-Date).AddSeconds(%d).ToString('HH:mm:ss'))"
+                    % (task_name, task_cmd, task_arg, delay_sec),
+                ],
+                shell=False,
+                startupinfo=startupinfo,
+            )
+
+            reply = {
+                "code": 0,
+                "content": "Restarting in %d seconds..."
+                % delay_sec,
+            }
+
+        except subprocess.CalledProcessError:
+            reply = {
+                "code": -1,
+                "cmd": "error",
+                "error": "Error creating restart task.",
+            }
+
+        return reply
+
+    def unix_firefox_restart(self, profile_dir, browser_cmd):
+        """ Restart Firefox on Unix like systems. """
+
+        reply = {}
+        ff_lock_name = "lock"
+
+        ff_bin_name = browser_cmd
+        ff_bin_path = '"%s"' % shutil.which(ff_bin_name)
+
+        ff_bin_dir = '"%s"' % str(
+            pathlib.Path(shutil.which(ff_bin_name)).parent
+        )
+
+        util = Utility()
+
+        if profile_dir == "auto":
+            ff_lock_path = ff_bin_path
+            ff_args = '"%s"' % ("-foreground")
+        else:
+            ff_lock_path = '"%s/%s"' % (profile_dir, ff_lock_name)
+            ff_args = '"%s","%s","%s"' % (
+                "-foreground",
+                "-profile",
+                profile_dir,
+            )
+
+        try:
+            restart_sh_name = "restart_firefox.sh"
+            restart_sh_dirname = self.get_native_dir()
+
+            restart_sh_path = "%s/%s" % (
+                restart_sh_dirname,
+                restart_sh_name,
+            )
+
+            restart_sh_content = """#!{unix_shell}
+
+debug={debug}
+
+export PATH="{path}"
+
+while readlink "{profile_dir}/lock"; do
+  sleep 1
+done
+
+sleep 1
+
+echo "[+] Executing pre-restart hook(s) ..."
+
+for hook in {native_dir}/{pre_restart_hook_prefix}*.sh; do
```

Comment:
Rather than executing every file in `native_dir` I feel like it'd be better to have users add the name of a script to their config and execute this one script. Would there be disadvantages to doing this?

Reply:
We are not actually executing every file in `native_dir`. We are only executing files with the following pattern - `pre_restart_hook-*.sh`. 

The `*` part is meant to vary based on the function that creates that hook. For example, currently we'd only have `pre_restart_hook-remove_firefox_prefs.sh`.

Another alternative could be to use the a folder named `hooks`, which is what Git does.

Thanks.

Reply:
Having a `hooks` folder with a single `pre_restart.sh` file inside indeed sounds better than adding something in the user's config.